File: app/src/main/java/com/lalmeeva/expense/utils/RxUtils.java
package com.lalmeeva.expense.utils;

import io.reactivex.Flowable;
import io.reactivex.FlowableTransformer;
import io.reactivex.Scheduler;

public class RxUtils {

    private final Scheduler mMainScheduler;
    private final Scheduler mBackgroundScheduler;

    public RxUtils(Scheduler mainScheduler, Scheduler backgroundScheduler) {
        mMainScheduler = mainScheduler;
        mBackgroundScheduler = backgroundScheduler;
    }

    private final FlowableTransformer flowableTransformer = new FlowableTransformer() {

        @Override
        public Flowable apply(Flowable upstream) {
            return upstream.subscribeOn(mBackgroundScheduler).observeOn(mMainScheduler);
        }
    };

    @SuppressWarnings("unchecked")
    public <T> FlowableTransformer<T, T> applySchedulersFlowable() {
        return (FlowableTransformer<T, T>) flowableTransformer;
    }

    public Scheduler getMainScheduler() {
        return mMainScheduler;
    }

    public Scheduler getBackgroundScheduler() {
        return mBackgroundScheduler;
    }

}

File: app/src/main/java/com/lalmeeva/expense/utils/RxUtils.kt
package com.lalmeeva.expense.utils

import com.lalmeeva.expense.base.view.ProgressableView
import io.reactivex.SingleTransformer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import timber.log.Timber

class RxUtils {
    companion object {

        fun <T> applySchedulersSingle(): SingleTransformer<T, T> {
            return SingleTransformer {
                it.subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
            }
        }

        fun <T> applyProgressSingle(view: ProgressableView?): SingleTransformer<T, T> {
            return SingleTransformer { it ->
                it.doOnSubscribe { view?.showProgress() }
                    .doOnError { Timber.e(it) }
                    .doFinally { view?.hideProgress() }
            }
        }
    }
}